packages/realm_dart/lib/src/logging.dart

```diff
@@ -0,0 +1,182 @@
+}
+
+class _RealmLogCategory extends RealmLogCategory {
+  _RealmLogCategory() : super._('Realm', null);
+
+  late final app = _LeafLogCategory('App', this);
+  late final sdk = _LeafLogCategory('SDK', this);
+  late final storage = _StorageLogCategory(this);
+  late final sync = _SyncLogCategory(this);
+}
+
+class _SyncLogCategory extends RealmLogCategory {
+  _SyncLogCategory(RealmLogCategory parent) : super._('Sync', parent);
+
+  late final client = _ClientLogCategory(this);
+  late final server = _LeafLogCategory('Server', this);
+}
+
+class _ClientLogCategory extends RealmLogCategory {
+  _ClientLogCategory(RealmLogCategory parent) : super._('Client', parent);
+
+  late final changeset = _LeafLogCategory('Changeset', this);
+  late final network = _LeafLogCategory('Network', this);
+  late final reset = _LeafLogCategory('Reset', this);
+  late final session = _LeafLogCategory('Session', this);
+}
+
+class _StorageLogCategory extends RealmLogCategory {
+  _StorageLogCategory(RealmLogCategory parent) : super._('Storage', parent);
+
+  late final notification = _LeafLogCategory('Notification', this);
+  late final object = _LeafLogCategory('Object', this);
+  late final query = _LeafLogCategory('Query', this);
+  late final transaction = _LeafLogCategory('Transaction', this);
+}
+
+/// Specifies the criticality level above which messages will be logged
+/// by the default sync client logger.
+/// {@category Realm}
+enum RealmLogLevel {
+  /// Log everything. This will seriously harm the performance of the
+  /// sync client and should never be used in production scenarios.
+  all(Level.ALL),
+
+  /// A version of [debug] that allows for very high volume output.
+  /// This may seriously affect the performance of the sync client.
+  trace(Level.FINEST),
+
+  /// Reveal information that can aid debugging, no longer paying
+  /// attention to efficiency.
+  debug(Level.FINER),
+
+  /// Same as [info], but prioritize completeness over minimalism.
+  detail(Level.FINE),
+
+  /// Log operational sync client messages, but in a minimalist fashion to
+  /// avoid general overhead from logging and to keep volume down.
+  info(Level.INFO),
+
+  /// Log errors and warnings.
+  warn(Level.WARNING),
+
+  /// Log errors only.
+  error(Level.SEVERE),
+
+  /// Log only fatal errors.
+  fatal(Level.SHOUT),
+
+  /// Turn off logging.
+  off(Level.OFF),
+  ;
+
+  static const logToValues = [
+    RealmLogLevel.trace,
+    RealmLogLevel.debug,
+    RealmLogLevel.detail,
+    RealmLogLevel.info,
+    RealmLogLevel.warn,
+    RealmLogLevel.error,
+    RealmLogLevel.fatal,
+  ];
+
+  final Level level;
+  const RealmLogLevel(this.level);
+
+  static final _valuesMap = {for (var value in RealmLogLevel.values) value.index: value};
+  factory RealmLogLevel.fromInt(int code) => _valuesMap[code]!;
+}
+
+/// A record of a log message from the Realm SDK.
+typedef RealmLogRecord = ({RealmLogCategory category, RealmLogLevel level, String message});
```

Comment:
Similarly to `RealmLogCategory`, can we make this a `LogRecord` or will it conflict with built-in types?

Reply:
See discussion about `RealmLogCategory`.